pstoreds/addr_book.go

```diff
@@ -309,57 +452,72 @@ func newTTLManager(parent context.Context, d ds.Datastore, c *lru.ARCCache, tick
 }
 
 // To be called by TTL manager's coroutine only.
-func (mgr *ttlmanager) tick() {
+func (mgr *ttlManager) tick() {
 	mgr.Lock()
 	defer mgr.Unlock()
 
 	now := time.Now()
-	batch, err := mgr.ds.Batch()
-	if err != nil {
-		log.Error(err)
+	var toDel []ds.Key
+	for key, entry := range mgr.entries {
+		if entry.ExpiresAt.After(now) {
+			continue
+		}
+		toDel = append(toDel, key)
+	}
+
+	if len(toDel) == 0 {
 		return
 	}
-	for key, entry := range mgr.entries {
-		if entry.ExpiresAt.Before(now) {
-			if err := batch.Delete(key); err != nil {
-				log.Error(err)
-			} else {
-				mgr.cache.Remove(key)
-			}
-			delete(mgr.entries, key)
+
+	txn := mgr.ds.NewTransaction(false)
+	defer txn.Discard()
+
+	for _, key := range toDel {
+		if err := txn.Delete(key); err != nil {
+			log.Error("failed to delete TTL key: %v, cause: %v", key.String(), err)
+			break
 		}
+		mgr.cache.Remove(key.Parent().Name())
```

Comment:
let's remove all from cache once transaction commits for safety

Reply:
My instinct tells me we want to mark the entries as "dirty" (i.e. evict them) as soon as they're marked for deletion. 

Otherwise we incur in a race between the txn commit and the eviction from cache (which would actually be another loop, so even more racy).

Reply:
i guess it's not problematic to have to read through the cache again on a failed txn. might even be better to remove from cache *before* removing from datastore now that i think of it. either way

Reply:
@bigs Yeah, I believe this logic does remove from cache _before_ removing from the datastore, as it precedes the commit. Another alternative is to reiterate through `toDel` _after doing all `tx.Delete()`s and before the commit._ That would ensure we evict all caches only if the individual deletes were successful and before the update is seen by the db. However, I wouldn't dwell on it because this TTL logic is going away soon.